modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java

```diff
@@ -230,12 +234,33 @@ public void start() {
                 plan = prepareSingle(qry0, pctx);
             }
 
+            checkQueryType(plan.type(), isQuery);
+
             cursors.add(executePlan(UUID.randomUUID(), pctx, plan));
         }
 
         return cursors;
     }
 
+    /**
+     * Check expected statement type (when it is set by JDBC) and given statement type.
+     *
+     * @param qryType Query type.
+     * @param isQuery {@code null} if not applicable, {@code true} for select queries,
+     *                           otherwise (DML/DDL queries) {@code false}.
+     */
+    private static void checkQueryType(Type qryType, Boolean isQuery) {
+        if (isQuery == null || qryType == Type.EXPLAIN) {
+            return;
+        }
+        if (isQuery && qryType != Type.QUERY) {
+            throw new IgniteInternalException("Given statement type does not match that declared by JDBC driver");
```

Comment:
Why does the error that is thrown from common executor module, refers to JDBC driver?

Reply:
I refactored error handling and moved all JDBC related messages to JdbcQueryEventHandler. Thnx.